### ISSUE TITLE
autogen.sh interfering with configure script

### ISSUE THREAD

**Issue by lava** [jemalloc/jemalloc#984]
I was going to write an issue about the `--with-jemalloc-prefix` option being broken, but while doing so I actually solved my problem. I'm not compltely sure if this is a bug, but since it took me quite some time to figure out maybe others will 

Starting from a fresh checkout:

    $ ./autogen.sh
    $ mkdir build && cd build
    $ ../configure --with-jemalloc-prefix=foo_
    [...]
    JEMALLOC_PREFIX: foo_
    JEMALLOC_PRIVATE_NAMESPACE: je_
    [...]
    $ make all && make install DESTDIR=foo
    $ nm -a foo/usr/local/lib/libjemalloc.so.2 | grep ' T '
    000000000000a340 T aligned_alloc
    000000000000a840 T calloc
    000000000000e230 T dallocx
    0000000000051efc T _fini
    0000000000004990 T free
    00000000000036e8 T _init
    000000000000f960 T mallctl
    000000000000fd50 T mallctlbymib
    000000000000fb70 T mallctlnametomib
    00000000000082a0 T malloc
    000000000000ff60 T malloc_stats_print
    00000000000101a0 T malloc_usable_size
    000000000000b5b0 T mallocx
    0000000000006930 T memalign
    000000000000f4d0 T nallocx
    0000000000009e50 T posix_memalign
    000000000000d0f0 T rallocx
    0000000000008c10 T realloc
    000000000000e010 T sallocx
    000000000000ea70 T sdallocx
    000000000000b1a0 T valloc
    000000000000dc50 T xallocx

What seems to happen is that `./autogen.sh` is already calling configure, and then the `--with-jemalloc-prefix` option from autogen (i.e. empty prefix) takes precedence over the one passed to configure in the build directory.

This could probably be solved by not calling configure from autogen.sh, but I don't actually understand the purpose behind the `--enable-autogen` option it enables, so maybe this would break something.

**Comment by interwq** [comment]
Does it work if you do "./autogen.sh --with-jemalloc-prefix=foo_" and drop the "./configure"? I don't think you need to do both. It looks the generated config file in source dir took precedence.

**Comment by lava** [comment]
Yes, this would of course also work, although it produces an in-tree build wich is usually not desired. In fact, once the issue is understood it can be easily avoided by just calling `autoconf` direcly instead of `autogen.sh`.

But this requires one to first understand that the call to `autogen.sh` will permanently change any future out-of-tree builds (I think it's because the autogenerated headers in the source directory come earlier in the include path than the one from the build directory) which is actually a pretty subtle thing to notice becaues all of the generated files are hidden by `.gitignore`.

**Comment by interwq** [comment]
@lava: thanks for the additional info. 

@davidtgoldblatt : is it easy to change the order of the #include paths? I just realized this also bit me a while ago (source dir config took place first).

**Comment by davidtgoldblatt** [comment]
Any reason not to just kill autogen.sh outright?

**Comment by jasone** [comment]
I use ```./autogen.sh ...``` instead of ```autoconf; ./configure ...```.  Once upon a time ```autogen.sh``` scripts were a common thing.

**Comment by lava** [comment]
I think the usage of `autogen.sh`-scripts is still reasonably common (although most are called `bootstrap` due to https://www.gnu.org/software/automake/faq/autotools-faq.html#What-does-_002e_002fbootstrap-or-_002e_002fautogen_002esh-do_003f ).

However, they usually just call `aclocal`, `autoheader`, `autoconf`, etc. as needed; this is the only project I can recall where `configure` is run as part of this chain.

**Comment by dragon512** [comment]
Came across this issue as I am having some bug trying to build with autoconf in a out of source setup which should work. I noticed the this build does not define a makefile.am file like it "should", but instead makefile.in

ideally the setup should be like 
```
autoreconf -i
./configure <options as needed>
make all ...
```

either way the out of source build breaks with:
```
...
/usr/bin/install: cannot stat 'doc/jemalloc.html': No such file or directory
Makefile:455: recipe for target 'install_doc_html' failed
```

hopefully a good cmake build setup will be added soon for linux
